File: algorithms/0739-daily-temperatures/solution.py
# Daily Temperatures - monotonic stack of indices waiting for a warmer day.
# Time O(n), Space O(n).

class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        n = len(temperatures)
        answer = [0] * n
        stack: list[int] = []
        for i, t in enumerate(temperatures):
            while stack and temperatures[stack[-1]] < t:
                j = stack.pop()
                answer[j] = i - j
            stack.append(i)
        return answer

File: algorithms/0739-daily-temperatures/Solution.kt
// Daily Temperatures - monotonic stack of indices waiting for a warmer day.
// Time O(n), Space O(n).

class Solution {
    fun dailyTemperatures(temperatures: IntArray): IntArray {
        val n = temperatures.size
        val answer = IntArray(n)
        val stack = ArrayDeque<Int>()
        for (i in 0 until n) {
            while (stack.isNotEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                val j = stack.pop()
                answer[j] = i - j
            }
            stack.push(i)
        }
        return answer
    }
}
